Add configurable progress and session info indicators

amitschang opened the pull request:
### Description
Configuration options are added to allow override of the spark statement progress indicator and for the html table of
session info upon startup. Both come with defaults that implement the same function as was done prior to this change, so
there would be no impact in case options are not set.

In my use-case, I wanted to create a indicator that sits at the top of the page and includes a link to the UI, and thus can be seen no matter where in the page your are focused. I think there are many other potential uses, such as users wanting to do simpler things like setting the width of the widget, or choosing a different widget for progress, etc.

*Testing*: Manually tested with notebook both without the new options and confirming the existing behavior and with overridden classes to customize both progress and startup display. Existing unit tests pass

*Note*: my editor is set to be opinionated about trailing white-space, so the PR includes some white-space fixes. I can certainly prune those out if it is not desired to include unrelated cleanup. Otherwise, the diff can be viewed with white-space hidden.

### Checklist
- [x] Wrote a description of my changes above 
- [x] Added a bullet point for my changes to the top of the `CHANGELOG.md` file
- [x] Added or modified unit tests to reflect my changes
- [x] Manually tested with a notebook
- [x] If adding a feature, there is an example notebook and/or documentation in the `README.md` file

devstein reviewed (commented):
Thanks for contributing and the helpful PR description @amitschang! I'll try to review this week.

devstein commented on sparkmagic/sparkmagic/utils/progress.py:
```diff
@@ -0,0 +1,32 @@
+    def close(self):
+        pass
+
+class defaultProgressIndicator(ProgressIndicator):
```
Two minor changes:

1. Please change to `TitleCase`
```suggestion
class DefaultProgressIndicator(ProgressIndicator):
```

2. Give it a more meaningful name, one idea would be

```suggestion
class HorizontalProgressBarIndicator(ProgressIndicator):
```

devstein commented on sparkmagic/sparkmagic/utils/startupinfo.py:
```diff
@@ -0,0 +1,20 @@
+    def display_info(self):
+        pass
+
+class defaultStartupInfoDisplay(StartupInfoDisplay):
```
Same naming comments 
```suggestion
class BasicStartupInfoDisplay(StartupInfoDisplay):
```

devstein commented on sparkmagic/sparkmagic/utils/configuration.py:
```diff
@@ -271,6 +271,13 @@ def kerberos_auth_configuration():
         "mutual_authentication": REQUIRED
     }
 
+@_with_override
+def progress_indicator_class():
+    return 'sparkmagic.utils.progress.defaultProgressIndicator'
+
+@_with_override
+def startup_info_display_class():
+    return 'sparkmagic.utils.startupinfo.defaultStartupInfoDisplay'
```
Reminder to change names here as well

devstein replied:
Can you add unit tests that will tell us if these paths point to a class that doesn't exist? This will be helpful for future refactors

devstein commented on sparkmagic/sparkmagic/utils/startupinfo.py:
```diff
@@ -0,0 +1,20 @@
+from sparkmagic.utils.utils import get_sessions_info_html
+
+class StartupInfoDisplay:
+    def __init__(self, ipython_display, sessions_info, current_session_id):
+        self.ipython_display = ipython_display
+        self.sessions_info = sessions_info
+        self.current_session_id = current_session_id
+
+    def write_msg(self, msg):
+        pass
+
+    def display_info(self):
+        pass
```
I found it confusing that `defaultProgressIndicator` calls `display` in its `init` function, but the `defaultStartupInfoDisplay` has `display_info` method that is called after initialization.

Can we consolidate on one pattern? What do you think?

devstein reviewed (changes requested):
Overall looks good. Just need to fix some issue around naming and class design

devstein commented:
@amitschang Merge conflicts from #758 an easy way to fix is

```shell
git remote add upstream git://github.com/jupyter-incubator/sparkmagic.git
git fetch upstream
# choose local changes over remotes
git merge upstream/master -X ours
# run black on all files
black .
```

amitschang commented:
Thanks @devstein! I will carve out some time to work on requested changes probably early next week, cheers :)

amitschang commented:
@devstein, I've addressed all your comments I believe. Please let me know what you think. I've also merged in the lastest upstream and ran black. There are still a small number of whitespace fixes left from my changes, for non-py files - I hope they are OK left in. The tests were added to the configuration tests since they have to to with default configuration and check whether the defaults are subclasses of the appropriate indicator classes (even though otherwise there is no attempt at enforcing this - which I think is probably OK).

amitschang commented:
a bit of a bump on this. Still something you think could be included?
